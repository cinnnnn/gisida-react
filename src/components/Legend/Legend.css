.legend {
  position: absolute;
  bottom: 40px;
  right: 20px;
  width: 300px;
  background-color: #f7f6ef;
  z-index: 1000;
  color: #555;
  font-size: 11px;
  line-height: 1.2;
}


.legend[data-visible="false"] {
    display: none;
  }

.legend .legend-row {
  padding: 15px;
  position: relative;
  box-shadow: 0 1px 7px rgba(0, 0, 0, 0.2) 
}

legend-row b {
    display: inline-block;
    padding-right: 14px;
  }

legend-row > span {
  padding-right: 30px;
  display: inline-block;
}