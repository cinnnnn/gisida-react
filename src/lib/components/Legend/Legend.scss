.legend {
  position: absolute;
  bottom: 40px;
  right: 20px;
  width: 300px;
  background-color: #f7f6ef;
  z-index: 1000;
  color: #555;
  font-size: 11px;
  line-height: 1.2;
}

.legend.legend-row {
  padding: 15px;
  position: relative;
  box-shadow: 0 1px 7px rgba(0, 0, 0, 0.2) 
}

.legend.legend-row b {
    display: inline-block;
    padding-right: 14px;
  }

.legend .legend-row > span {
  padding-right: 30px;
  display: inline-block;
}

.legend[data-visible="false"] {
    display: none;
  }


.legend-fill {
  height: 10px;
  background: #eee;
  margin: 3px 0;
}

.legend-fill ul {
    padding: 0;
    list-style-type: none;
}
.legend-fill li {
    width: 30px;
    float: left;
    display: block;
    height: 10px;
    color: #fff;
    opacity: 0.8;
}

.legend-fill.legend-label {
  height: 17px;
}

.legend-fill.legend-label li {
    text-align: center;
    height: 17px;
    padding-top: 2px;
}


.legend-shapes {
  position: relative;
}

.legend-shapes .circle-sm {
    height: 6px;
    width: 6px;
    margin: 0 10px 0 0;
    border-radius: 50%;
    background: #eee;
    display: inline-block;
}

.legend-shapes .circle-md {
    height: 9px;
    width: 9px;
    margin: 0 10px 0 0;
    border-radius: 50%;
    background: #eee;
    display: inline-block;
}

.legend-shapes .circle-lg {
    height: 12px;
    width: 12px;
    margin: 0 10px 0 0;
    border-radius: 50%;
    background: #eee;
    display: inline-block;
}

.legend-shapes .triangle-stroked-11 {
    width: 0;
    height: 0;
    margin: 0 10px 0 0;
    border-left: 4px solid transparent;
    border-right: 4px solid transparent;
    border-bottom: 8px solid white;
    display: inline-block;
}

.legend-shapes .triangle-15 {
    width: 0;
    height: 0;
    margin: 0 10px 0 0;
    border-left: 6px solid transparent;
    border-right: 6px solid transparent;
    border-bottom: 12px solid white;
    display: inline-block;
}