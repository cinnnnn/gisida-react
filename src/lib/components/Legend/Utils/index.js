import React from 'react';
export const buildTimestamp = (layerObj) => {
    if (
        layerObj['timestamp'] &&
        Array.isArray(layerObj.source.data.features || layerObj.source.data) &&
        (layerObj.source.data.features || layerObj.source.data).length > 1
      ) {
        return (layerObj.source.data.features || layerObj.source.data)
          .map(d => (d.properties && d.properties[layerObj['timestamp']]) || d[layerObj['timestamp']])
          .sort()
          .reverse()[0];
      }
      return null;
}


export const buildHyperLink = (layerObj) => {
    return (<a href={layerObj['legendHyperlink']} target="_blank" data-link="resourceLink">
                {layerObj['legendHyperlinkText'] ?  <h5 data-link="resourceLink">{layerObj['legendHyperlinkText']}</h5> : null}
                <span data-link="resourceLink" className='glyphicon glyphicon-file' />
            </a>);
}

export const buildDescription = (layerObj) => {
    return <div className='description'>
                <span className='glyphicon glyphicon-info-sign' />
                <p>{layerObj['legendDescription']}</p>
		   </div>;
}

export const combinedLinks = (legendDescription, legendHyperlink) => {
    if (legendHyperlink || legendDescription) {
        return (<div className='combinedLinks'>
                    {legendDescription}&nbsp;&nbsp;
                    {legendHyperlink} 
                </div>);
            } else {
                return null
            }
}