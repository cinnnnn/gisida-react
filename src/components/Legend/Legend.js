import React from 'react';
import './Legend.css';

const mapId = '01';

const Legend = () => (<div className={`legend ${mapId}`} />);

export default Legend;
