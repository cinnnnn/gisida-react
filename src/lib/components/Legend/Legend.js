/* eslint-disable no-loop-func */
import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Actions, formatNum, hexToRgbA } from 'gisida';
import { buildLayersObj } from '../../utils';
import Parser from 'html-react-parser';
import './Legend.scss';

const mapStateToProps = (state, ownProps) => {
  return {
    layerObj: state.MAP.layers[state.MAP.activeLayerId],
    layersData: buildLayersObj(state.MAP.layers),
    MAP: state.MAP,
    primaryLayer: state.MAP.primaryLayer,
  }
}

export class Legend extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      setPrimary: false,
    };
  }

  onUpdatePrimaryLayer(e) {
    e.preventDefault();
    const dispatch = this.props.dispatch;
    const targetLayer = e.currentTarget.getAttribute('data-layer');
    // dispatch primary layer id
    dispatch(Actions.updatePrimaryLayer(targetLayer));
  }

  componentWillReceiveProps(nextProps) {

  }

  render() {
    const mapId = '01';
    const { layerObj } = this.props;
    if (!layerObj) {
      return false;
    }
    const legendItems = [];

    let primaryLegend;
    let layer;

    for (let l = 0; l < this.props.layersData.length; l += 1) {
      layer = this.props.layersData[l];
      const circleLayerType = (layer && layer.credit && layer.type === 'circle' && !layer.categories.shape && layer.visible);
      const symbolLayer = (layer && layer.credit && layer.categories &&  layer.categories.shape && layer.type !== 'circle'); 
      const fillLayerNoBreaks = (layer && layer.credit && layer.categories && layer.categories.breaks === 'no');
      const fillLayerWithBreaks = (layer && layer.credit && layer.type !== 'chart' && layer.type !== 'circle');
      const activeLayerSelected =  this.props.primaryLayer === layer.id ? 'primary' : '';
      const lastSelected = this.props.layersData[this.props.layersData.length - 1].id === layer.id  ? 'primary' : '';

      const boundaries = ["region-boundaries", "district-boundaries"];
      const isBoundaries = boundaries.indexOf(layer.id) !== -1;

      let background = [];

      if (layerObj.id === layer.id) {
        if (circleLayerType) {
          primaryLegend = (
        <div
          id={`legend-${layer.id}-${mapId}`}
          className={`legend-shapes legend-row ${activeLayerSelected}`}
          data-layer={`${layer.id}`}
          onClick={(e) => this.onUpdatePrimaryLayer(e)}
          key={l}
        >
          <b>
            {layer.label}
          </b>
          <div className="legend-symbols">
          <span
            className="circle-sm"
            style={{ background: layer.categories.color }}
          >
          </span>
          <span
            className="circle-md"
            style={{ background: layer.categories.color }}
          >
          </span>
          <span
            className="circle-lg"
            style={{ background: layer.categories.color }}
          >
          </span>
          </div>
          <span>{layer.credit}</span>
        </div>);
        }
        if (fillLayerNoBreaks) {
          layer.categories.color.forEach((color, index) => {
            const fillWidth = (100 / layer.categories.color.length).toString();
            background.push((
              <li
                key={index}
                style={{ background: color, width: fillWidth + '%' }}>
                {layer.categories.label[index]}
              </li>
            ));
          });

          const legendClass = layer.categories ? 'legend-label' : '';

          primaryLegend = (
            <div
              id={`legend-${layer.id}-${mapId}`}
              className={`legend-row ${activeLayerSelected}`}
              data-layer={`${layer.id}`}
              onClick={(e) => this.onUpdatePrimaryLayer(e)}
              key={l}
            >
              <b>
                {layer.label}
              </b>
              <div className={`legend-fill ${legendClass}`}>
                <ul>
                  {background}
                </ul>
              </div>
              <span>
                {Parser(layer.credit)}
              </span>
            </div>
          );
        }
        if (fillLayerWithBreaks && layerObj.stops && layer.stops) {
          const { stopsData, breaks, colors, Data } = layerObj;

          const dataValues = Data.map(values => values[layer.property]);
          const colorLegend = [...new Set(stopsData.map(stop => stop[1]))];
          const legendSuffix = layer.categories.suffix ? layer.categories.suffix : '';

          if (colorLegend.includes('transparent') && !(colors).includes('transparent')) {
            colors.splice(0, 0, 'transparent');
            breaks.splice(1, 0, breaks[0]);
          }

          colorLegend.forEach((color, index) => {
            const firstVal = breaks[index - 1] !== undefined ? breaks[index - 1] : 0;
            const lastVal = color === colorLegend[colorLegend.length - 1] || breaks[index] === undefined ? Math.max(...dataValues) : breaks[index];
            background.push((
              <li
                key={index}
                className={`background-block-${layer.id}-${mapId}`}
                data-tooltip={`${formatNum(firstVal, 1)}-${formatNum(lastVal, 1)}${legendSuffix}`}
                style={{ background: hexToRgbA(color, 0.9).toString(), width: (100 / colorLegend.length) + '%' }}
              >
              </li>
            ));
          });

          primaryLegend = (
            <div
            id={`legend-${layer.id}-${mapId}`}
            className={`legend-row ${activeLayerSelected || lastSelected}`}
            data-layer={`${layer.id}`}
            key={l}
            onClick={(e) => this.onUpdatePrimaryLayer(e)}
          >
            <b>
              {layer.label}
            </b>
            <ul
              className="legend-limit"
              style={{ padding: '0% 0% 3% 0%' }}
            >
              <li
                id={`first-limit-${layer.id}`}
                className={`${mapId}`}
                style={{ position: 'absolute', listStyle: 'none', display: 'inline', left: '3%' }}
              >
                {0}
                {legendSuffix}
              </li>
              <li
                id={`last-limit-${layer.id}`}
                className={`${mapId}`}
                style={{ position: 'absolute', listStyle: 'none', display: 'inline', right: '3%' }}
              >
                {formatNum(Math.max(...dataValues), 1)}
                {legendSuffix}
              </li>
            </ul>
            <div
              className="legend-fill"
            >
              <ul
                id="legend-background"
              >
                {background}
              </ul>
            </div>
            <span>{Parser(layer.credit)}</span>
          </div>
          );

        }
        if (!isBoundaries && !symbolLayer && !circleLayerType && !fillLayerNoBreaks && !fillLayerWithBreaks) {
          primaryLegend = (
            <div
              id={`legend-${layer.id}-${mapId}`}
              className={`legend-row primary`}
              data-layer={`${layer.id}`}
              key={l}
            >
              <b>
                {layer.label}
              </b>
              <span>
                {layer.credit ? Parser(layer.credit) : ''}
              </span>
            </div>
          );
        }
        continue;
      }
      if (circleLayerType) {
        legendItems.unshift((
          <div
            id={`legend-${layer.id}-${mapId}`}
            className={`legend-shapes legend-row ${activeLayerSelected || lastSelected}`}
            data-layer={`${layer.id}`}
            key={l}
            onClick={(e) => this.onUpdatePrimaryLayer(e)}
          >
            <b>
              {layer.label}
            </b>
            <div className="legend-symbols">
              <span
                className="circle-sm"
                style={{ background: layer.categories.color }}
              >
              </span>
              <span
                className="circle-md"
                style={{ background: layer.categories.color }}
              >
              </span>
              <span
                className="circle-lg"
                style={{ background: layer.categories.color }}
              >
              </span>
            </div>
            <span>{layer.credit}</span>
          </div>
        ));
      } else if (symbolLayer) {
        layer.categories.color.forEach((color, index) => {
          const style = layer.categories.shape[index] === 'triangle-stroked-11' ||
          layer.categories.shape[index] === 'triangle-15' ?
          'border-bottom-color:' : 'background:';
          const styleString = `${style}: ${color}`;
          background += (
            <li
              className="layer-symbols"
              key={index}
            >
              <span
                className={`${layer.categories.shape[index]}`}
                style={{styleString}}
              />
              {layer.categories.label[index]}
            </li>
          );
        });

        legendItems.unshift((
          <div
            id={`legend-${layer.id}-${mapId}`}
            className={`legend-row ${activeLayerSelected || lastSelected}`}
            data-layer={`${layer.id}`}
            onClick={(e) => this.onUpdatePrimaryLayer(e)}
            key={l}
          >
            <b>
              {layer.label}
            </b>
            <div className="legend-shapes">
              <ul style={{ left: '0' }}>
                {background}
              </ul>
            </div>
            <span>
              {Parser(layer.credit)}
            </span>
          </div>
        ));

      } else if (fillLayerNoBreaks) {
        layer.categories.color.forEach((color, index) => {
          const fillWidth = (100 / layer.categories.color.length).toString();
          background.push((
            <li
              key={index}
              style={{ background: color, width: fillWidth + '%' }}>
              {layer.categories.label[index]}
            </li>
          ));
        });

        const legendClass = layer.categories ? 'legend-label' : '';

        legendItems.unshift((
          <div
            id={`legend-${layer.id}-${mapId}`}
            className={`legend-row ${activeLayerSelected || lastSelected}`}
            data-layer={`${layer.id}`}
            onClick={(e) => this.onUpdatePrimaryLayer(e)}
            key={l}
          >
            <b>
              {layer.label}
            </b>
            <div className={`legend-fill ${legendClass}`}>
              <ul>
                {background}
              </ul>
            </div>
            <span>
            {Parser(layer.credit)}
            </span>
          </div>
        ));
      } else if (fillLayerWithBreaks && layer.stops) {
        const { stopsData, breaks, colors, Data } = layer;

        const dataValues = Data.map(values => values[layer.property]);
        const colorLegend = [...new Set(stopsData.map(stop => stop[1]))];
        const legendSuffix = layer.categories.suffix ? layer.categories.suffix : '';

        if (colorLegend.includes('transparent') && !(colors).includes('transparent')) {
          colors.splice(0, 0, 'transparent');
          breaks.splice(1, 0, breaks[0]);
        }

        colorLegend.forEach((color, index) => {
          const firstVal = breaks[index - 1] !== undefined ? breaks[index - 1] : 0;
          const lastVal = color === colorLegend[colorLegend.length - 1] || breaks[index] === undefined ? Math.max(...dataValues) : breaks[index];
          background.push((
            <li
              key={index}
              className={`background-block-${layer.id}-${mapId}`}
              data-tooltip={`${formatNum(firstVal, 1)}-${formatNum(lastVal, 1)}${legendSuffix}`}
              style={{ background: hexToRgbA(color, 0.9).toString(), width: (100 / colorLegend.length) + '%' }}
            >
            </li>
          ));
        });

        legendItems.unshift((
          <div
            id={`legend-${layer.id}-${mapId}`}
            className={`legend-row ${activeLayerSelected}`}
            data-layer={`${layer.id}`}
            key={l}
            onClick={(e) => this.onUpdatePrimaryLayer(e)}
          >
            <b>
              {layer.label}
            </b>
            <ul
              className="legend-limit"
              style={{ padding: '0% 0% 3% 0%' }}
            >
              <li
                id={`first-limit-${layer.id}`}
                className={`${mapId}`}
                style={{ position: 'absolute', listStyle: 'none', display: 'inline', left: '3%' }}
              >
                {0}
                {legendSuffix}
              </li>
              <li
                id={`last-limit-${layer.id}`}
                className={`${mapId}`}
                style={{ position: 'absolute', listStyle: 'none', display: 'inline', right: '3%' }}
              >
                {formatNum(Math.max(...dataValues), 1)}
                {legendSuffix}
              </li>
            </ul>
            <div
              className="legend-fill"
            >
              <ul
                id="legend-background"
              >
                {background}
              </ul>
            </div>
            <span>{Parser(layer.credit)}</span>
          </div>
        ));
      }
      else {
        if (!isBoundaries && !symbolLayer && !circleLayerType && !fillLayerNoBreaks && !fillLayerWithBreaks) {
          legendItems.unshift((
            <div
              id={`legend-${layer.id}-${mapId}`}
              className={`legend-row`}
              data-layer={`${layer.id}`}
              key={l}
            >
              <b>
                {layer.label}
              </b>
              <span>
                {typeof layer.credit === "string" ? Parser(layer.credit) : null}
              </span>
            </div>
          ));
        }
      }
    }

    legendItems.unshift(primaryLegend);

    return (
      <div>
        <div className={`legend ${mapId}`}>
          {legendItems}
        </div>
      </div>
    );
  }
}

Legend.propTypes = {
  layerObj: PropTypes.objectOf(PropTypes.any).isRequired,
  layersData: PropTypes.arrayOf(PropTypes.any).isRequired,
  MAP: PropTypes.objectOf(PropTypes.any).isRequired,
  primaryLayer: PropTypes.string.isRequired,
};

export default connect(mapStateToProps)(Legend);
