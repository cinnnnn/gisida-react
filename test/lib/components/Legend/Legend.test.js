import React from 'react';
import Legend from '../../../../src/lib/components/Legend/Legend'
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';


const componentWrapper = shallow(
  <Legend />
);

describe('Legend', () => {
  it('component renderes correctly', () => {
    const json = toJson(componentWrapper)
    expect(json).toMatchSnapshot();
  });
});
