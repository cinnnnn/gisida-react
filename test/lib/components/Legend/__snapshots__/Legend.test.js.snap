// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Legend component renderes correctly 1`] = `
<div>
  <div
    className="legend 01"
    style={
      Object {
        "right": "20px",
      }
    }
  >
    <div
      className="legend-row"
      data-layer="undefined"
      id="legend-undefined-01"
      key="0"
    >
      <b />
      <span />
    </div>
  </div>
</div>
`;
